makefile

```diff
@@ -1,3 +1,53 @@
-proto:
-	@mkdir -p gopb
-	protoc --go_out=plugins=grpc:gopb *.proto
+# start project configuration
+buildDir := build
+pbDir := gopb
+# end project configuration
+
+# start environment setup
+gobin := go
+ifneq (,$(GOROOT))
+gobin := $(GOROOT)/bin/go
+endif
+
+goCache := $(GOCACHE)
+ifeq (,$(goCache))
+goCache := $(abspath $(buildDir)/.cache)
+endif
+goModCache := $(GOMODCACHE)
+ifeq (,$(goModCache))
+goModCache := $(abspath $(buildDir)/.mod-cache)
+endif
+
+ifneq ($(goCache),$(GOCACHE))
+export GOCACHE := $(goCache)
+endif
+ifneq ($(goModCache),$(GOMODCACHE))
+export GOMODCACHE := $(goModCache)
+endif
+# end environment setup
+
+# ensure the build directory exists, since most targets require it
+$(shell mkdir -p $(buildDir))
+
+# start basic development targets
+protocVersion := 3.6.1
+protocGenGoVersion := 1.3.2
+protoOS := $(shell uname -s | tr A-Z a-z)
+ifeq ($(protoOS),darwin)
+protoOS := osx
+endif
+protoOS := $(protoOS)-$(shell uname -m | tr A-Z a-z)
+$(buildDir)/protoc:
+	curl --retry 10 --retry-max-time 60 -L0 https://github.com/protocolbuffers/protobuf/releases/download/v$(protocVersion)/protoc-$(protocVersion)-$(protoOS).zip --output protoc.zip
+	unzip -q protoc.zip -d $(buildDir)/protoc
+	rm -f protoc.zip
+	GOBIN="$(abspath $(buildDir))" $(gobin) install github.com/golang/protobuf/protoc-gen-go@v$(protocGenGoVersion)
+proto: $(buildDir)/protoc
+	mkdir -p $(pbDir)
+	PATH="$(abspath $(buildDir)):$(PATH)" $(buildDir)/protoc/bin/protoc --go_out=plugins=grpc:$(pbDir) *.proto
+# end basic development targets
+
+# start cleanup targets
+clean:
+	rm -rf $(buildDir) $(pbDir)
```

Comment:
I would make a separate `clean-proto` for removing the protobuf file directory.

Reply:
And not the binaries?

Reply:
I consider `make clean` to mean "clean up artifacts created as a result of building and are not committed". Since the protobuf files are committed, make clean would actually change the state of the git repo.

Reply:
Honestly, you don't even need to remove the .pb.go files at all since they'll be regenerated if you type `make proto` anyways.

Reply:
That is true. Ok I will remove it entirely then.